### PR TITLE
MSCXXXX: Limit maximum number of events sent to an AS

### PR DESCRIPTION
**Pull request by Half-Shot**

### CODE REVIEW

File: proposals/3982-proposal-template.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+
+## Alternatives
+
+Potentially the homeserver and application service could negotiate a maximum event size instead of specifying
+one in the spec, however this isn't desirable because:
+
+ - This requires a two way communication between the Homeserver and AS, which presently isn't required.
+ - This potentially costs more time in negotiation than any efficiency losses from the hardcoded limit.
+ - Practically speaking, Synapse [has had a limit of 100](https://github.com/matrix-org/synapse/blob/develop/synapse/appservice/scheduler.py#L85-L86)
```
Comment (Half-Shot) [inline proposals/3982-proposal-template.md:58]: For other homeserver implementations:

- Dendrite has a limit of 50 https://github.com/matrix-org/dendrite/blob/main/appservice/consumers/roomserver.go#L84
- Conduit has a limit of 30 https://gitlab.com/famedly/conduit/-/blob/next/src/service/sending/mod.rs#L159 (I *think*, it's kinda hard to follow)

File: proposals/3982-proposal-template.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+### Maximum theoretical payload size
+
+Assuming the worst case, the payload should be no larger than:
+
+```
+(65536 * 100) + 13 = 6553613 bytes
+```
+
+where 113 accounts for the JSON structure around the `events` array.
```
Comment (clokep) [inline proposals/3982-proposal-template.md:37]: 13 or 113?

File: proposals/3982-proposal-template.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+The biggest loss here is that this effectively caps the number of events a single payload can contain,
+and so there is potential for a cap to limit how quickly events are sent to an AS. However, a 50 event
+limit is fairly high and a sensible [HTTP Keep-Alive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive)
+value will keep a HTTP connection open allowing for quicker streaming of events to the appservice.
+
+
+## Alternatives
+
+Potentially the homeserver and application service could negotiate a maximum event size instead of specifying
+one in the spec, however this isn't desirable because:
```
Comment (clokep) [inline proposals/3982-proposal-template.md:54]: Another reason this is not reasonable is that neither Synapse nor the application service know about any middle boxes between them, so neither of them can possibly know the maximum HTTP request size that is allowed without trial & error.

File: proposals/3982-proposal-template.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+## Proposal
+
+The [`PUT /_matrix/app/v1/transactions/{txnId}` endpoint](https://spec.matrix.org/v1.6/application-service-api/#put_matrixappv1transactionstxnid)
+MUST only allow up to `100` events in the `events` array in it's body.
+
+This spec follows the precedent set by the [Server-Server /transaction api](https://spec.matrix.org/v1.6/server-server-api/#put_matrixfederationv1sendtxnid)
+of setting size limits on transaction payloads.
+
+If the homeserver needs to send more than `100` events in a transaction, it should
+send them in a subsequent transaction.
```
Comment (clokep) [inline proposals/3982-proposal-template.md:23]: Should we also set a maximum EDU & to-device limit like Synapse does:

https://github.com/matrix-org/synapse/blob/827f198177c4cf547b9d2d1eed41411e945fc199/synapse/appservice/scheduler.py#L85-L92

Comment (clokep) [reply proposals/3982-proposal-template.md:23]: I guess this is really a concern [MSC2409](https://github.com/matrix-org/matrix-spec-proposals/pull/2409).

Comment (Half-Shot) [reply proposals/3982-proposal-template.md:23]: I've chosen to ignore that MSC while it's still in unstable land, though I'd expect that MSC to adopt a similar thing.

Comment (clokep) [reply proposals/3982-proposal-template.md:23]: 👍 Makes sense.
